certora/specs/CER-2-Operator/CER-68-Operator-authorization.spec

```diff
@@ -0,0 +1,142 @@
+import "../utils/IsMustRevertFunction.spec";
+
+methods {
+    function getOwnerOf(bytes19) external returns (address) envfree;
+    function getOperator(bytes19, address) external returns (uint256) envfree;
+    function getAddressPrefix(address) external returns (bytes19) envfree;
+    function haveCommonOwner(address account, address otherAccount) external returns (bool) envfree;
+}
+
+/**
+ * Check that `setOperator(addressPrefix, ...)` can only be called if msg.sender
+ * is the owner of the address prefix. Technically, we check that
+ * - msg.sender is from the prefix itself and thus a plausible owner
+ * - msg.sender is stored as the owner after the function call
+ * - the owner before the call was either 0 (not set) or msg.sender already
+ * - the bitset is set as it should be.
+ */
+rule onlyOwnerCanCallSetOperator() {
+    env e;
+
+    bytes19 addressPrefix;
+    address operator;
+    uint256 operatorBitField;
+
+    address ownerBefore = getOwnerOf(addressPrefix);
+
+    // if msg.sender is the currentContract, it means we are within permit() and
+    // we need to use executionContext.getOnBehalfOfAccount() instead.
+    address actualCaller = e.msg.sender;
+    if (e.msg.sender == currentContract) {
+        actualCaller = getExecutionContextOnBehalfOfAccount(e);
+    }
+
+    // call the setOperator() method.
+    setOperator(e, addressPrefix, operator, operatorBitField);
+
+    // sender is from the prefix itself and thus plausible to be the owner
+    assert(getAddressPrefix(actualCaller) == addressPrefix);
+    // the owner before the call was either not set or actualCaller already
+    assert(ownerBefore == 0 || ownerBefore == actualCaller);
+    // sender is stored as the owner of the address prefix
+    assert(actualCaller == getOwnerOf(addressPrefix));
+    // make sure the right bitfield was set
+    assert(getOperator(addressPrefix, operator) == operatorBitField);
+}
+
+rule onlyOwnerOrOperatorCanCallSetAccountOperator() {
+    env e;
+    address account;
+    address operator;
+    bool authorized;
+
+    // if msg.sender is the currentContract, it means we are within permit() and
+    // we need to use executionContext.getOnBehalfOfAccount() instead.
+    address actualCaller = e.msg.sender;
+    if (e.msg.sender == currentContract) {
+        actualCaller = getExecutionContextOnBehalfOfAccount(e);
+    }
+
+    // call the setAccountOperator method.
+    setAccountOperator(e, account, operator, authorized);
+
+    address owner = haveCommonOwner(account, actualCaller) ? actualCaller : getAccountOwner(e, account);
+
+    // Since setAccountOperator did not revert, the actualCaller
+    // must either be the owner or operator
+    assert(actualCaller == owner || actualCaller == operator);
+
+}
+
+// a copy of the internal ownerLookup
+ghost mapping(bytes19 => address) ownerLookupGhost {
```

Comment:
is this ghost used?

Reply:
It is not used aside from your suggestion about how to fix the case for batch/call which I could not get working. So I can delete this for now.

Reply:
Used now!